src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs

```diff
@@ -103,33 +99,58 @@ private static bool IsDeadProducer(IStreamProducerExtension producer)
 
         public async Task<ISet<PubSubSubscriptionState>> RegisterProducer(StreamId streamId, IStreamProducerExtension streamProducer)
         {
-            if (!IsActiveProducer(streamProducer))
-                throw new ArgumentException(String.Format("Trying to register non active IStreamProducerExtension: {0}", streamProducer.ToString()), "streamProducer");
-            
-            RemoveDeadProducers();
-            
-            var publisherState = new PubSubPublisherState(streamId, streamProducer);
-            State.Producers.Add(publisherState);
             counterProducersAdded.Increment();
-            counterProducersTotal.Increment();
-            LogPubSubCounts("RegisterProducer {0}", streamProducer);
-            await WriteStateAsync();
+            if (!IsActiveProducer(streamProducer))
+                throw new ArgumentException($"Trying to register non active IStreamProducerExtension: {streamProducer}", "streamProducer");
+
+            try
+            {
+                int producersRemoved = RemoveDeadProducers();
+
+                var publisherState = new PubSubPublisherState(streamId, streamProducer);
+                State.Producers.Add(publisherState);
+                LogPubSubCounts("RegisterProducer {0}", streamProducer);
+                await WriteStateAsync();
+                counterProducersTotal.DecrementBy(producersRemoved);
+                counterProducersTotal.Increment();
+            }
+            catch (Exception exc)
+            {
+                logger.Error(ErrorCode.Stream_RegisterProducerFailed, $"Failed to register a stream producer.  Stream: {streamId}, Producer: {streamProducer}, Exception: {exc}");
+                // Corrupted state, deactivate grain.
+                DeactivateOnIdle();
+                throw;
+            }
             return State.Consumers.Where(c => !c.IsFaulted).ToSet();
         }
 
         public async Task UnregisterProducer(StreamId streamId, IStreamProducerExtension streamProducer)
         {
-            int numRemoved = State.Producers.RemoveWhere(s => s.Equals(streamId, streamProducer));
             counterProducersRemoved.Increment();
-            counterProducersTotal.DecrementBy(numRemoved);
-            LogPubSubCounts("UnregisterProducer {0} NumRemoved={1}", streamProducer, numRemoved);
+            try
```

Comment:
not necessarily an issue, but I'm curious why registering a producer triggers RemoveDeadProducers, but unregistering one doesn't